src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/ProjectSystem/RazorProjectInfoEndpointPublisher.cs

```diff
@@ -0,0 +1,205 @@
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Razor.ProjectSystem;
+using Microsoft.CodeAnalysis.Razor.Protocol;
+using Microsoft.CodeAnalysis.Razor.Utilities;
+using Microsoft.CodeAnalysis.Razor.Workspaces.Protocol.ProjectSystem;
+using Microsoft.VisualStudio.LanguageServer.ContainedLanguage;
+
+namespace Microsoft.VisualStudio.LanguageServerClient.Razor.ProjectSystem;
+
+/// <summary>
+/// Publishes project data (including TagHelper info) discovered OOB to the server via LSP notification
+/// instead of old method of writing a project configuration bin file
+/// </summary>
+[Shared]
+[Export(typeof(RazorProjectInfoEndpointPublisher))]
+internal partial class RazorProjectInfoEndpointPublisher : IDisposable
+{
+    private readonly LSPRequestInvoker _requestInvoker;
+    private readonly IProjectSnapshotManager _projectManager;
+
+    private readonly AsyncBatchingWorkQueue<(IProjectSnapshot Project, bool Removal)> _workQueue;
+    private readonly CancellationTokenSource _disposeTokenSource;
+    private bool _active;
+
+    [ImportingConstructor]
+    public RazorProjectInfoEndpointPublisher(
+        LSPRequestInvoker requestInvoker,
+        IProjectSnapshotManager projectManager
+    )
+    {
+        _requestInvoker = requestInvoker;
+        _projectManager = projectManager;
+
+        _disposeTokenSource = new ();
+        _workQueue = new(
+            EnqueueDelay,
+            ProcessBatchAsync,
+            _disposeTokenSource.Token);
+    }
+
+    public void Dispose()
+    {
+        _disposeTokenSource.Cancel();
+        _disposeTokenSource.Dispose();
+    }
+
+    // internal for testing
+    // Delay between publishes to prevent bursts of changes yet still be responsive to changes.
+    internal TimeSpan EnqueueDelay { get; set; } = TimeSpan.FromMilliseconds(250);
+
+    public void StartSending()
+    {
+        _active = true;
+
+        _projectManager.Changed += ProjectManager_Changed;
+
+        var projects = _projectManager.GetProjects();
+        foreach (var project in projects)
+        {
+            if (ProjectWorkspacePublishable(project))
+            {
+                // Don't enqueue project addition as those are unlikely to come through in large batches.
+                // Also ensures that we won't get project removal go through the queue without project addition.
+                ImmediatePublish(project, _disposeTokenSource.Token);
+            }
+        }
+    }
+
+    private void ProjectManager_Changed(object sender, ProjectChangeEventArgs args)
+    {
+        // Don't do any work if the solution is closing
+        if (args.SolutionIsClosing)
+        {
+            return;
+        }
+
+        switch (args.Kind)
+        {
+            case ProjectChangeKind.ProjectChanged:
+            case ProjectChangeKind.DocumentRemoved:
+            case ProjectChangeKind.DocumentAdded:
+                if (!ProjectWorkspacePublishable(args.Newer))
```

Comment:
Consider using our `AssumeNotNull` extension method so that you don't need the `!` below or deal with a nullable `IProjectSnapshot` in `ProjectWorkspacePublishable`.